More pause options

Henrybk opened the pull request:
This pull request add the possibility of pausing automacro checking and/or the running macro, as two different things. I also added the corresponding console commands. Needs testing.

Henrybk committed:
v1

Henrybk committed:
Added check in clear_queue

Henrybk committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into More-pause-options

Henrybk committed:
Fixed bug in AI_pre hooking

allanon commented on eventMacro.pl:
```diff
@@ -92,10 +92,13 @@ sub commandHandler {
-			"eventMacro stop: stop current macro\n".
-			"eventMacro pause: interrupt current macro\n".
-			"eventMacro resume: resume interrupted macro\n".
+			"eventMacro macro_status: shows current macro status\n".
```
These names are too long to type. It's fine to assume people have an IDE and will be able to autocomplete long variable names, but these are going to be typed by end users. They should be short and easy to remember. I vote to keep the original names for the `macro` set (ie, `status`, `stop`, `pause`, `resume`).

Henrybk replied:
I like 'macro_status|stop|pause|resume' but I also think that the automacro commands got way bigger than I would like them to be. Maybe 'automacro_pause|resume|status'?

allanon commented on eventMacro/Core.pm:
```diff
@@ -70,24 +68,22 @@ sub clean_hooks {
+sub set_automacro_checking_status {
+	my ($self, $status) = @_;
+	
+	if ($self->{Automacros_Checking_Status} == CHECKING_AUTOMACROS) {
```
What happens if `$self->{Automacros_Checking_Status} == $status`?

Henrybk replied:
There was a return for this case in the previous commit, I removed it by accident.

allanon commented on eventMacro/Core.pm:
```diff
@@ -70,24 +68,22 @@ sub clean_hooks {
+		Plugins::delHook($self->{AI_pre_Hook_Handle});
+		$self->{AI_pre_Hook_Handle} = undef;
+	} elsif ($status == CHECKING_AUTOMACROS) {
+		$self->{AI_pre_Hook_Handle} = Plugins::addHook( 'AI_pre', sub { $self->AI_pre_checker(); }, undef );
```
The trailing `undef` is unnecessary.

allanon commented on eventMacro.pl:
```diff
@@ -134,15 +137,75 @@ sub commandHandler {
+		my $status = $eventMacro->get_automacro_checking_status();
+		if ($status == 0) {
+			message "Automacros won't be checked anymore.\n";
+			$eventMacro->set_automacro_checking_status(2);
```
Should we use the names instead? Eg, `eventMacro::Core::PAUSED_BY_USER`.

Henrybk replied:
Good idea

allanon commented on eventMacro.pl:
```diff
@@ -92,10 +92,13 @@ sub commandHandler {
+			"eventMacro macro_pause: pauses current macro execution\n".
+			"eventMacro macro_resume: resumes interrupted macro\n".
+			"eventMacro automacro_checking_status: shows current automacro checking status\n".
+			"eventMacro automacro_checking_pause: pauses automacro checking\n".
```
Maybe make this `eventMacro disable automacros`? That sounds a little better (to me).

Henrybk committed:
Fixed bug with command called macros

Fixed bug with command called macros, also changed some functions calls
to use Core constants.

allanon commented on eventMacro.pl:
```diff
@@ -188,10 +251,14 @@ sub commandHandler {
-			if ( defined $opt->{exclusive} )   { $eventMacro->{Macro_Runner}->interruptible( 0 ); }
+			if ( defined $opt->{exclusive} )   {
+				$eventMacro->{Macro_Runner}->interruptible( 0 );
+				message "[eventMacro] Calling uninterruptible macro '".$arg."'. Automacro checking will be paused until it ends.\n";
```
Should this be a `debug` instead?

Henrybk committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into More-pause-options

Henrybk committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into More-pause-options

Henrybk committed:
Commands changed and hood deletion added

Henrybk committed:
merge with master

Henrybk committed:
Changes in macro inheritance and AI_pre

Moved 'eventMacro::Core::set_automacro_checking_status()' call to
'eventMacro::Runner::interruptible'.
Now all macros subcalls will inherit it's father exclusive value, if the
called macro sets it's own exclusive value it will only last until
eventMacro::Runner returns to the caller macro, then the caller macro
exclusive value will take place again.

Henrybk commented:
Here is a macro to test the added features in inheritance: http://pastebin.com/FKiYrwzy

To better see the changes it's a good idea to add a message or debug to eventMacro::Core::AI_pre_checker

Henrybk committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into More-pause-options

# Conflicts:
#	eventMacro/Core.pm
#	eventMacro/Runner.pm

Henrybk committed:
Bugfix in event only automacros
